### PR TITLE
Fix bugs and add tests

### PR DESCRIPTION
**Pull request by BehindTheMath**
This PR has a bunch of various bug fixes for the code and tests, and many new tests. It's based on #144, so that needs to be merged first.

@robinnorth I cherry-picked your commit from the `fix/missing-tests` branch (Add tests for `lib/util/noop.js` - 5e7341d).

### CODE REVIEW

**Comment by robinnorth** [comment]
@BehindTheMath nice one, that's a substantial PR! I'll take a proper look at it as soon as I have time.

**Comment by BehindTheMath** [comment]
Thanks. I didn't originally plan on it being so big, but I was aiming for as close to 100% coverage as I could, and I kept adding more tests and finding more issues.

If it's too big, I can break it up into smaller PRs.

**Comment by BehindTheMath** [comment]
@robinnorth Do you think you'll have time to look this over, or should I just merge it? I would love to have a second pair of eyes on it, but I know you said you've been busy lately.

**Comment by robinnorth** [comment]
Sorry, I've been very tardy in reviewing this for you. I have indeed been busy, but I'll make some time to check it out later and get back to you.

**Comment by BehindTheMath** [comment]
Thank you, I appreciate it.

File: lib/proto/parse-element.js [hunk 1]
```diff
@@ -2,14 +2,14 @@ module.exports = function(el) {
   switch (el.tagName.toLowerCase()) {
     case "a":
       // only attach link if el does not already have link attached
-      if (!el.hasAttribute("data-pjax-click-state")) {
+      if (!el.hasAttribute("data-pjax-state")) {
```
Comment (robinnorth) [inline lib/proto/parse-element.js]: This, and line 12 could be abstracted to use `var attrState = "data-pjax-state"`, as in `attach-link`

File: tests/lib/proto/attach-form.js [hunk 2]
```diff
@@ -29,50 +27,57 @@ tape("test attach form prototype method", function(t) {
...
 
   t.end()
 })
 
 tape("test attach form preventDefaulted events", function(t) {
-  var callbacked = false
+  var loadUrlCalled = false
   var form = document.createElement("form")
 
+  // This needs to be before the call to attachFormk()
```
Comment (robinnorth) [inline tests/lib/proto/attach-form.js]: Is `attachFormk` a typo?

Comment (BehindTheMath) [reply tests/lib/proto/attach-form.js]: Yup

File: lib/proto/parse-element.js (hunk 1 above)
Comment (BehindTheMath) [reply lib/proto/parse-element.js]: True.

**Comment by BehindTheMath** [comment]
I deleted the `fix/missing-tests` branch, since I cherry-picked the only commit from there into this PR, and we're close to 100% coverage.

**Comment by robinnorth** [comment]
Cool, thanks!

**Pull request merged**